app/validators/personal_key_validator.rb

```diff
@@ -0,0 +1,44 @@
+module PersonalKeyValidator
+  extend ActiveSupport::Concern
+
+  included do
+    include ActiveModel::Validations::Callbacks
+
+    before_validation :normalized_code
+
+    attr_accessor :personal_key
+    attr_reader :user
+
+    validate :valid_personal_key?
+  end
+
+  private
+
+  def normalized_code
```

Comment:
I think `normalized_code` is a misleading name, since it's normalizing the `personal_key` value not the `code` value